### PR TITLE
Fixes issues with unlabelled groups on aromatic nitrogens

### PR DESCRIPTION
**Pull request by ptosco**
Make sure that unlabelled R-groups on aromatic nitrogens have correct valence and formal charge to avoid kekulization failures.
I have seen the problem on substituted pyrrole-like nitrogens (which became `[nH](*)` rather than `n(*)` as they should.
I have also seen the problem on pyridinum-like systems (which stayed as `n(*)` rather than becoming` [n+](*)`).
I have also noticed that the JSON tests were failing due to a spurious R-group label showing up on the core but not actually being used, so I also fixed the JSON tests.

### CODE REVIEW

**Comment by greglandrum** [comment]
@ptosco: The "spurious" R group labels which were in the original output were user-provided labels from the core, so I don't think it's necessarily wrong that they were making it through. 

I also don't have a problem with what the output of this PR is doing - only including R groups which actually match - but I want to be sure that this change in behavior was intentional and that @jones-gareth and @bp-kelley agree that it's correct.

**Comment by ptosco** [comment]
@greglandrum OK, now I have checked the SDF file and I see your point. I think I also want them to be preserved. I'll edit my PR accordingly.

**Comment by ptosco** [comment]
OK, I looked a bit deeper into this.

If we want to preserve user-specified labels, which could be desirable, I think they need to appear also among R-groups, as either `H` or empty, as appropriate, otherwise it would be inconsistent.

In this respect, I think the JSON was inconsistent also before my proposed change, as the R-groups showed up in the core only.

Also, there are a number of tests:
* test rgroup decomp ring matching
* test rgroup decomp full ring dummy core
* testSDFGRoupMultiCoreNoneShouldMatch
* Test core with matching at any atom

where cores do have user-specified labels which should be preserved in the output for both the core and R-groups if we wish this behavior.

@greglandrum @bp-kelley @jones-gareth Please let me know your thoughts before I make this change.

**Comment by greglandrum** [comment]
I don't understand why you think the JSON was inconsistent beforehand. It looks fine to me.
Why do you think it's inconsistent?

**Comment by ptosco** [comment]
I guess I have been staring at RGD too long.
I thought that if a label shows up in `Core` (because it was user-defined) it should also show up as an R-group, even though it is empty or `H` for all molecules.
Otherwise, it's probably OK to remove it also from the `Core`.
Or not?

**Comment by ptosco** [comment]
Anyway, I believe all of these tests:
* test rgroup decomp ring matching
* test rgroup decomp full ring dummy core
* testSDFGRoupMultiCoreNoneShouldMatch
* Test core with matching at any atom
Present the same situation: user-specified label in the `Core`, which currently is missing in test expected output.
I am happy to change the code behavior and to revert the JSON as it was, but then I think we need to fix those 4 tests as well: if we want to preserve user-specified labels, this should happen consistently.
Please be patient if I am totally wrong and tell me why I am wrong. I might eventually understand.

**Comment by jones-gareth** [comment]
> 
> 
> I guess I have been staring at RGD too long.
> I thought that if a label shows up in `Core` (because it was user-defined) it should also show up as an R-group, even though it is empty or `H` for all molecules.
> Otherwise, it's probably OK to remove it also from the `Core`.
> Or not?

It's weird - the longer you look at RGD the less clear it is.

As part of #3767 I think we agreed that we wouldn't show unused labeled R groups on the core and that we would remove all groups containing all H or nothing from the output.

I can see that from a user's perspective that it may be confusing to have the R group they have drawn absent from the core output, so we may wish to revisit.  Personally, I am good with either the R group being present or not.

While it might appear to be inconsistent to have an R group present in the core, but missing from the columns output (because it contains all H or nothing), from a practical point of view this is fine.  RGD results are going to end up in a UI and nobody is going to want to see a column of all Hs- but they *may* want to see the missing R group in the core.

**Comment by jones-gareth** [comment]
> 
> 
> Anyway, I believe all of these tests:
> 
>     * test rgroup decomp ring matching
> 
>     * test rgroup decomp full ring dummy core
> 
>     * testSDFGRoupMultiCoreNoneShouldMatch
> 
>     * Test core with matching at any atom
> 
> 
> present the same situation: user-specified label in the `Core`, which currently is missing in test expected output.
> I am happy to change the code behavior and to revert the JSON as it was, but then I think we need to fix those 4 tests as well: if we want to preserve user-specified labels, this should happen consistently.
> Please be patient if I am totally wrong and tell me why I am wrong. I might eventually understand.

You are right about those tests- I had to change those specs when merging your PR against #3767, when the unused labeled R groups were removed from the core.

**Comment by ptosco** [comment]
@jones-gareth OK, that makes perfect sense. I am fine with empty/H R-groups not showing up.
We only need whether we want to keep user labels in the `Core` or not.
In case we do, I'll update the tests accordingly. @greglandrum @bp-kelley could you give your opinion?
I have a slight preference for keeping user labels.

Side note: if we choose to keep all user-defined R-labels, we might have occasional kekulization failures when we have things like thiophene with an R-group sprouting from the S. That may happen with things like `*1***[*:1]1` (`testRingMatching3`): being user-labelled, the R-group will not be removed, and the kekulizer will fail because sulfur can be tetravalent but in this case it is only trivalent and this will confuse the kekulizer - I have seen this in the past few days while playing with labels. This does not happen with furan as the O is only divalent and the kekulizer tolerates that.

**Comment by greglandrum** [comment]
After a brief offline conversation with @ptosco I can see the argument for having the core which appears in the output represent what actually matched, so I'm fine with this change.
Will review tomorrow morning.

File: Code/GraphMol/RGroupDecomposition/RGroupCore.h [hunk 1]
```diff
@@ -60,88 +60,95 @@ struct RCore {
...
+                                            const ROMol &mol,
+                                            const MatchVectType &match) const {
+    auto coreReplacedAtoms = boost::make_shared<RWMol>(*core);
+    hasCoreDummies = false;
     for (const auto &p : match) {
-      auto atom = coreReplacedDummies->getAtomWithIdx(p.first);
-      if (isAnyAtomWithMultipleNeighborsOrNotUserRLabel(*atom)) {
-        hasDummies = true;
+      auto atom = coreReplacedAtoms->getAtomWithIdx(p.first);
+      hasCoreDummies |= (atom->getAtomicNum() == 0);
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupCore.h:73]: I think it's more readable this way:
```suggestion
      if(atom->getAtomicNum() == 0){
        hasCoreDummies = true;
      }
```

**Review by greglandrum — APPROVED** [review]
LGTM aside from the one code-formatting nit.
I'm fine merging as-is if you don't feel like changing that.

**Comment by greglandrum** [comment]
As a random aside, while looking at this I realized that we can also use the RGD code with linker groups by using cores which have multiple fragments:
```

In [3]: core = Chem.MolFromSmarts('C1CC1.C1CCC1')

In [4]: from rdkit.Chem import rdRGroupDecomposition

In [5]: rdRGroupDecomposition.RGroupDecompose([core],ms,asRows=True,asSmiles=True)
Out[5]: 
([{'Core': 'C1C([*:3])C1[*:4].C1CC([*:2])C1[*:1]',
   'R1': 'C(C[*:1])C[*:4]',
   'R2': 'F[*:2]',
   'R3': 'Cl[*:3]',
   'R4': 'C(C[*:1])C[*:4]'},
  {'Core': 'C1C([*:3])C1[*:4].C1CC([*:2])C1[*:1]',
   'R1': 'C(C[*:1])C[*:4]',
   'R2': 'F[*:2]',
   'R3': 'O[*:3]',
   'R4': 'C(C[*:1])C[*:4]'}],
 [])
```
that's pretty cool. :-)

File: Code/GraphMol/RGroupDecomposition/RGroupCore.h (hunk 1 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupCore.h]: No probs, done. I was 99.9% sure that wouldn't pass the `greg-clang-format` filter, but I thought I'd try. I think it's a good thing also to check your attention level and keep you snappy.

Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupCore.h]: > No probs, done. I was 99.9% sure that wouldn't pass the `greg-clang-format` filter, but I thought I'd try. I think it's a good thing also to check your attention level and keep you snappy.

Thanks for giving me that opportunity! ;-)

**Review by greglandrum — APPROVED** [review]

**Comment by bp-kelley** [comment]
@ptosco I had a review that apparently I never sent.  My internet has been super flakey up in the mountains.  Thanks @ptosco @jones-gareth @greglandrum for all the effort on this one.

**Pull request merged**